# Real Estate Price Prediction

User want to know the price of the house of unit area depends on 6 parameter as mentioned below.
Transaction date, house age, distance to the nearest MRT station, number of convenience stores in the living circle, geographic coordinate latitude, geographic coordinate longitude.

- Attribute Information:

- The inputs are as follows
- X1=the transaction date (for example, 2013.250=2013 March, 2013.500=2013 June, etc.)
- X2=the house age (unit: year)
- X3=the distance to the nearest MRT station (unit: meter)
- X4=the number of convenience stores in the living circle on foot (integer)
- X5=the geographic coordinate, latitude. (unit: degree)
- X6=the geographic coordinate, longitude. (unit: degree)

The output is as follow
Y= house price of unit area (10000 New Taiwan Dollar/Ping, where Ping is a local unit, 1 Ping = 3.3 meter squared)

# Importing required libraries for the project

In [24]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

- Loading real estate dataset from git hub account

In [25]:
Realdf = pd.read_csv('https://raw.githubusercontent.com/Manju410/MLPractice/main/Real_Estate_Price_Prediction/Real_estate.csv')

In [26]:
Realdf.head()

No  X1 transaction date  ...  X6 longitude  Y house price of unit area
0   1             2012.917  ...     121.54024                        37.9
1   2             2012.917  ...     121.53951                        42.2
2   3             2013.583  ...     121.54391                        47.3
3   4             2013.500  ...     121.54391                        54.8
4   5             2012.833  ...     121.54245                        43.1

[5 rows x 8 columns]

- Number of rows and columns in the dataset

In [27]:
Realdf.shape

(414, 8)

- There are 414 rows and 8 columns in the above dataset

- Information about dataset like datatype,count etc

In [28]:
Realdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 414 entries, 0 to 413
Data columns (total 8 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   No                                      414 non-null    int64  
 1   X1 transaction date                     414 non-null    float64
 2   X2 house age                            414 non-null    float64
 3   X3 distance to the nearest MRT station  414 non-null    float64
 4   X4 number of convenience stores         414 non-null    int64  
 5   X5 latitude                             414 non-null    float64
 6   X6 longitude                            414 non-null    float64
 7   Y house price of unit area              414 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 26.0 KB


# Summary of above output
- Above dataset contains 8 columns.
- Two columns are integer datatype and Six columns are float datatype.
- Above dataset doesnot have any null values or empty values.
- Above dataset have 414 etries total

- Checking null values in dataset

In [29]:
Realdf.isna().sum()

No                                        0
X1 transaction date                       0
X2 house age                              0
X3 distance to the nearest MRT station    0
X4 number of convenience stores           0
X5 latitude                               0
X6 longitude                              0
Y house price of unit area                0
dtype: int64

- Removing No column

Real estae dataset contain No column which is no related to outcom or lable variable so we are droping that column.

In [30]:
#Realdf1 = Realdf[Realdf.columns[1:]][:]
Realdf.drop('No', axis=1, inplace=True)
Realdf.head()

X1 transaction date  X2 house age  ...  X6 longitude  Y house price of unit area
0             2012.917          32.0  ...     121.54024                        37.9
1             2012.917          19.5  ...     121.53951                        42.2
2             2013.583          13.3  ...     121.54391                        47.3
3             2013.500          13.3  ...     121.54391                        54.8
4             2012.833           5.0  ...     121.54245                        43.1

[5 rows x 7 columns]

-X1=the transaction date (for example, 2013.250=2013 March, 2013.500=2013 June, etc.)

https://archive.ics.uci.edu/ml/datasets/Real+estate+valuation+data+set

from the description we have below

- We are Checking Unique values from transaction date to know how many Year and month variation.

In [31]:
date = Realdf['X1 transaction date'].unique().tolist()
date.sort()
date

[2012.6670000000001,
 2012.75,
 2012.8329999999999,
 2012.9170000000001,
 2013.0,
 2013.0829999999999,
 2013.1670000000001,
 2013.25,
 2013.3329999999999,
 2013.4170000000001,
 2013.5,
 2013.5829999999999]

- As per above transaction date and description given for transaction date
max 1000 count equally divided into 12 parts that is 12 months so each month carrying 83.33 

In [32]:
1000/12

83.33333333333333

- numpy has function called linspace which will select the random numbers from mentioned start and end index with mentioned stepsize

In [33]:
np.linspace(1,10,10,endpoint=False)

array([1. , 1.9, 2.8, 3.7, 4.6, 5.5, 6.4, 7.3, 8.2, 9.1])

- as per transaction date column we have min 2012.667 to max 2013.583 with stepsize of 12 values or months so we are applying linspace function to get 12 different values which will match above transaction date.
 
- here 2012 and 2013 are years
- and numbers after decimal point are months as per below calculated

In [34]:
np.linspace(2012.667,2013.583,12)

array([2012.667     , 2012.75027273, 2012.83354545, 2012.91681818,
       2013.00009091, 2013.08336364, 2013.16663636, 2013.24990909,
       2013.33318182, 2013.41645455, 2013.49972727, 2013.583     ])

- so as per above max value calculation each month corresponds to 83.3 value so min vlaue 2012.667 which represent 2012.Aug and max value 2013.583 represent 2013.July
- 667 = August : (83.3*8) & 583 = July : (83.3*7)
- as per above calculation am spliting 12 months start from August

In [35]:
months = 'Aug Sep Oct Nov Dec Jan Feb Mar Apr May Jun Jul'.split()
months

['Aug',
 'Sep',
 'Oct',
 'Nov',
 'Dec',
 'Jan',
 'Feb',
 'Mar',
 'Apr',
 'May',
 'Jun',
 'Jul']

- Replacing transaction date column with proper year month format using manually created months list and values in transaction date column

In [36]:
trndate_dict = {date[i]:str(date[i]).split('.')[0]+'-'+months[i] for i in range(len(date))}
trndate_dict

{2012.6670000000001: '2012-Aug',
 2012.75: '2012-Sep',
 2012.8329999999999: '2012-Oct',
 2012.9170000000001: '2012-Nov',
 2013.0: '2013-Dec',
 2013.0829999999999: '2013-Jan',
 2013.1670000000001: '2013-Feb',
 2013.25: '2013-Mar',
 2013.3329999999999: '2013-Apr',
 2013.4170000000001: '2013-May',
 2013.5: '2013-Jun',
 2013.5829999999999: '2013-Jul'}

- Getting transaction date column values in Year and month format by applying above analysis of date time calculation

In [37]:
Realdf['X1 transaction date'] = [trndate_dict[x] for x in Realdf['X1 transaction date']]

In [38]:
#Realdf['X1 transaction date'] = Realdf['X1 transaction date'].apply(datetransform)

- converting Transaction date column to datetime formate

In [39]:
Realdf['X1 transaction date']= pd.to_datetime(Realdf['X1 transaction date'])

In [40]:
Realdf.head()

X1 transaction date  X2 house age  ...  X6 longitude  Y house price of unit area
0          2012-11-01          32.0  ...     121.54024                        37.9
1          2012-11-01          19.5  ...     121.53951                        42.2
2          2013-07-01          13.3  ...     121.54391                        47.3
3          2013-06-01          13.3  ...     121.54391                        54.8
4          2012-10-01           5.0  ...     121.54245                        43.1

[5 rows x 7 columns]

- Checking information about dataset to check transaction date column datatype changed from int to datetime

In [41]:
Realdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 414 entries, 0 to 413
Data columns (total 7 columns):
 #   Column                                  Non-Null Count  Dtype         
---  ------                                  --------------  -----         
 0   X1 transaction date                     414 non-null    datetime64[ns]
 1   X2 house age                            414 non-null    float64       
 2   X3 distance to the nearest MRT station  414 non-null    float64       
 3   X4 number of convenience stores         414 non-null    int64         
 4   X5 latitude                             414 non-null    float64       
 5   X6 longitude                            414 non-null    float64       
 6   Y house price of unit area              414 non-null    float64       
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 22.8 KB


In [42]:
#abc= datetransform('2012.9170000000001')
#pd.to_datetime(abc)

- LinearRegression model doesnt accept datetime format so converting to numerical

In [43]:
import datetime as dt
Realdf['X1 transaction date']=Realdf['X1 transaction date'].map(dt.datetime.toordinal)

- toordinal used to convert datetime format to numerical ordinal values.
- importing datetime module and applying map function to transaction date

In [44]:
Realdf['X1 transaction date'][0]

734808

In [45]:
Realdf.head()

X1 transaction date  X2 house age  ...  X6 longitude  Y house price of unit area
0               734808          32.0  ...     121.54024                        37.9
1               734808          19.5  ...     121.53951                        42.2
2               735050          13.3  ...     121.54391                        47.3
3               735020          13.3  ...     121.54391                        54.8
4               734777           5.0  ...     121.54245                        43.1

[5 rows x 7 columns]

In [46]:
from google.colab import  drive

In [47]:
drive.mount('/drive')

Mounted at /drive


In [49]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [52]:
Realdf.to_csv('/content/drive/MyDrive/RealEstatCleanUp.csv', index=False)